File: src/external/tauri/commands/kanban.ts
import { invokeCommand } from '../core/invoker';

import type { Kanban } from '~/models/kanban';

export const kanbanCommand = {
  list: async (workspace: { slugName: string }) => {
    return await invokeCommand<Kanban[]>('list_kanbans', {
      workspace_slug_name: workspace.slugName,
    });
  },

  create: async (params: { workspaceSlugName: string; name: string }) => {
    return await invokeCommand<Kanban>('create_kanban', {
      workspace_slug_name: params.workspaceSlugName,
      name: params.name,
    });
  },

  updateStatusRoles: async (params: { workspaceSlugName: string; id: number; defaultStatusId?: number | null; focusStatusId?: number | null }) => {
    await invokeCommand('update_kanban_status_roles', {
      workspace_slug_name: params.workspaceSlugName,
      id: params.id,
      default_status_id: params.defaultStatusId ?? null,
      focus_status_id: params.focusStatusId ?? null,
    });
  },

  delete: async (params: { workspaceSlugName: string; id: number }) => {
    await invokeCommand('delete_kanban', {
      workspace_slug_name: params.workspaceSlugName,
      id: params.id,
    });
  },
};

File: src-tauri/src/commands/kanban.rs
use crate::database::get_conn;
use crate::models::kanban::Kanban;
use crate::repositories::{KanbanRepository, WorkspaceRepository};
use serde::Deserialize;
use tauri::command;

#[derive(Deserialize)]
pub struct ListKanbansArgs {
    pub workspace_slug_name: String,
}

#[command]
pub fn list_kanbans(args: ListKanbansArgs) -> Result<Vec<Kanban>, String> {
    let conn = get_conn().map_err(|e| e.to_string())?;

    let workspace = WorkspaceRepository::find_by_slug(&conn, &args.workspace_slug_name)
        .map_err(|e| e.to_string())?
        .ok_or_else(|| format!("Workspace not found for slug: {}", args.workspace_slug_name))?;

    let kanban = KanbanRepository::ensure_global_status_board(&conn, workspace.id)
        .map_err(|e| e.to_string())?;

    Ok(vec![kanban])
}

#[derive(Deserialize)]
pub struct CreateKanbanArgs {
    pub workspace_slug_name: String,
    pub name: String,
}

#[command]
pub fn create_kanban(args: CreateKanbanArgs) -> Result<Kanban, String> {
    let conn = get_conn().map_err(|e| e.to_string())?;

    let workspace = WorkspaceRepository::find_by_slug(&conn, &args.workspace_slug_name)
        .map_err(|e| e.to_string())?
        .ok_or_else(|| format!("Workspace not found for slug: {}", args.workspace_slug_name))?;

    KanbanRepository::create(&conn, workspace.id, &args.name).map_err(|e| e.to_string())
}

#[derive(Deserialize)]
pub struct UpdateKanbanStatusRolesArgs {
    pub workspace_slug_name: String,
    pub id: i32,
    pub default_status_id: Option<i32>,
    pub focus_status_id: Option<i32>,
}

#[command]
pub fn update_kanban_status_roles(args: UpdateKanbanStatusRolesArgs) -> Result<(), String> {
    let conn = get_conn().map_err(|e| e.to_string())?;

    let workspace = WorkspaceRepository::find_by_slug(&conn, &args.workspace_slug_name)
        .map_err(|e| e.to_string())?
        .ok_or_else(|| format!("Workspace not found for slug: {}", args.workspace_slug_name))?;

    let updated = KanbanRepository::update_status_roles(
        &conn,
        workspace.id,
        args.id,
        args.default_status_id,
        args.focus_status_id,
    )
    .map_err(|e| e.to_string())?;

    if !updated {
        return Err(format!("Kanban not found: {}", args.id));
    }

    Ok(())
}

#[derive(Deserialize)]
pub struct DeleteKanbanArgs {
    pub workspace_slug_name: String,
    pub id: i32,
}

#[command]
pub fn delete_kanban(args: DeleteKanbanArgs) -> Result<(), String> {
    let conn = get_conn().map_err(|e| e.to_string())?;

    let workspace = WorkspaceRepository::find_by_slug(&conn, &args.workspace_slug_name)
        .map_err(|e| e.to_string())?
        .ok_or_else(|| format!("Workspace not found for slug: {}", args.workspace_slug_name))?;

    let deleted =
        KanbanRepository::delete(&conn, workspace.id, args.id).map_err(|e| e.to_string())?;

    if !deleted {
        return Err(format!("Kanban not found: {}", args.id));
    }

    Ok(())
}
